File: neon_tetris.js
// neon_tetris.js
const blessed = require('blessed');
const screen = blessed.screen({
    smartCSR: true,
    title: 'Neon Tetris'
});

const WIDTH = 10;
const HEIGHT = 20;
const SHAPES = [
    [[1,1,1,1]],
    [[1,1],[1,1]],
    [[0,1,0],[1,1,1]],
    [[1,0,0],[1,1,1]],
    [[0,0,1],[1,1,1]],
    [[0,1,1],[1,1,0]],
    [[1,1,0],[0,1,1]]
];
const COLORS = ['cyan', 'yellow', 'magenta', 'green', 'red', 'blue', 'white'];

const board = Array.from({length: HEIGHT}, () => Array(WIDTH).fill(0));
let score = 0, lines = 0, level = 1, fallTime = 500, lastFall = Date.now();
let gameOver = false, paused = false;
let currentPiece = null, nextPiece = null, pieceX, pieceY;

function newPiece() {
    const idx = Math.floor(Math.random() * SHAPES.length);
    return { shape: SHAPES[idx], color: COLORS[idx] };
}

function spawnPiece() {
    currentPiece = nextPiece || newPiece();
    nextPiece = newPiece();
    pieceX = Math.floor(WIDTH/2 - currentPiece.shape[0].length/2);
    pieceY = 0;
    if (collision(currentPiece, pieceX, pieceY)) {
        gameOver = true;
    }
}

function collision(piece, x, y) {
    for (let row=0; row<piece.shape.length; row++) {
        for (let col=0; col<piece.shape[row].length; col++) {
            if (piece.shape[row][col]) {
                const bx = x + col;
                const by = y + row;
                if (bx < 0 || bx >= WIDTH || by >= HEIGHT) return true;
                if (by >= 0 && board[by][bx]) return true;
            }
        }
    }
    return false;
}

function lockPiece() {
    for (let row=0; row<currentPiece.shape.length; row++) {
        for (let col=0; col<currentPiece.shape[row].length; col++) {
            if (currentPiece.shape[row][col]) {
                const by = pieceY + row;
                const bx = pieceX + col;
                if (by >= 0) board[by][bx] = 1;
            }
        }
    }
    clearLines();
    spawnPiece();
}

function clearLines() {
    let cleared = 0;
    for (let y=HEIGHT-1; y>=0; y--) {
        if (board[y].every(v => v !== 0)) {
            board.splice(y, 1);
            board.unshift(Array(WIDTH).fill(0));
            cleared++;
            y++; // recheck
        }
    }
    if (cleared) {
        lines += cleared;
        score += cleared * 100;
        level = Math.floor(lines / 10) + 1;
        fallTime = Math.max(100, 500 - (level-1) * 30);
    }
}

function movePiece(dx, dy) {
    const nx = pieceX + dx;
    const ny = pieceY + dy;
    if (!collision(currentPiece, nx, ny)) {
        pieceX = nx;
        pieceY = ny;
        return true;
    }
    if (dy === 1) lockPiece();
    return false;
}

function rotatePiece() {
    const shape = currentPiece.shape;
    const rotated = shape[0].map((_, idx) => shape.map(row => row[idx]).reverse());
    const newPiece = { shape: rotated, color: currentPiece.color };
    if (!collision(newPiece, pieceX, pieceY)) {
        currentPiece = newPiece;
    } else {
        // wallkick
        for (const dx of [-1, 1]) {
            if (!collision(newPiece, pieceX+dx, pieceY)) {
                pieceX += dx;
                currentPiece = newPiece;
                break;
            }
        }
    }
}

function drop() {
    while (!collision(currentPiece, pieceX, pieceY+1)) pieceY++;
    lockPiece();
}

// UI
const box = blessed.box({
    top: 'center',
    left: 'center',
    width: 40,
    height: 24,
    border: { type: 'line' },
    style: { border: { fg: 'cyan' } }
});

const content = blessed.box({
    parent: box,
    top: 1,
    left: 1,
    width: 38,
    height: 22,
    content: ''
});
screen.append(box);

function render() {
    let out = '';
    // рамка поля
    out += '┌' + '─'.repeat(WIDTH) + '┐\n';
    for (let y=0; y<HEIGHT; y++) {
        out += '│';
        for (let x=0; x<WIDTH; x++) {
            let ch = ' ';
            let color = 'white';
            if (board[y][x]) {
                ch = '█'; color = COLORS[(x+y) % 6];
            }
            // current
            for (let row=0; row<currentPiece.shape.length; row++) {
                for (let col=0; col<currentPiece.shape[row].length; col++) {
                    if (currentPiece.shape[row][col]) {
                        const by = pieceY + row;
                        const bx = pieceX + col;
                        if (by === y && bx === x) {
                            ch = '█'; color = currentPiece.color;
                        }
                    }
                }
            }
            // цвет
            if (ch !== ' ') {
                out += `{${color}-fg}${ch}{/}`;
            } else {
                out += ' ';
            }
        }
        out += '│\n';
    }
    out += '└' + '─'.repeat(WIDTH) + '┘\n';
    out += `Score: ${score}  Level: ${level}  Lines: ${lines}\n`;
    out += `Next: `;
    // next piece preview
    for (let row=0; row<nextPiece.shape.length; row++) {
        for (let col=0; col<nextPiece.shape[row].length; col++) {
            if (nextPiece.shape[row][col]) {
                out += `{${nextPiece.color}-fg}█{/}`;
            } else out += ' ';
        }
        out += ' ';
    }
    if (paused) out += '\nPAUSED';
    if (gameOver) out += '\nGAME OVER';
    content.setContent(out);
    screen.render();
}

// key handling
screen.key(['q', 'C-c'], () => process.exit(0));
screen.key(['p'], () => { paused = !paused; });
screen.key(['left'], () => { if (!paused) movePiece(-1,0); });
screen.key(['right'], () => { if (!paused) movePiece(1,0); });
screen.key(['down'], () => { if (!paused) movePiece(0,1); });
screen.key(['up'], () => { if (!paused) rotatePiece(); });
screen.key(['space'], () => { if (!paused) drop(); });

// инициализация
spawnPiece();

// game loop
setInterval(() => {
    if (!paused && !gameOver) {
        const now = Date.now();
        if (now - lastFall > fallTime) {
            movePiece(0, 1);
            lastFall = now;
        }
    }
    render();
}, 30);

screen.render();

File: neon_tetris.rs
// neon_tetris.rs
use crossterm::{
    cursor, event::{self, Event, KeyCode, KeyEventKind},
    execute, terminal,
};
use std::io::stdout;
use std::time::{Duration, Instant};
use rand::Rng;

const WIDTH: usize = 10;
const HEIGHT: usize = 20;

const SHAPES: [[[u8; 4]; 4]; 7] = [
    // I
    [[0,0,0,0],[1,1,1,1],[0,0,0,0],[0,0,0,0]],
    // O
    [[1,1],[1,1]],
    // T
    [[0,1,0],[1,1,1]],
    // L
    [[1,0,0],[1,1,1]],
    // J
    [[0,0,1],[1,1,1]],
    // S
    [[0,1,1],[1,1,0]],
    // Z
    [[1,1,0],[0,1,1]],
];

struct Piece {
    shape: Vec<Vec<u8>>,
    color: u16, // цвет для терминала
}

impl Piece {
    fn new(idx: usize) -> Self {
        let raw = SHAPES[idx];
        let mut shape = Vec::new();
        for row in raw.iter() {
            let mut r = Vec::new();
            for &v in row {
                if v != 0 { r.push(1); } else { r.push(0); }
            }
            if !r.is_empty() { shape.push(r); }
        }
        let color = 33 + (idx % 6) as u16; // ANSI коды цветов
        Piece { shape, color }
    }
}

struct Tetris {
    board: [[u8; WIDTH]; HEIGHT],
    score: u32,
    lines: u32,
    level: u32,
    fall_time: u64,
    last_fall: Instant,
    game_over: bool,
    paused: bool,
    current: Piece,
    next: Piece,
    piece_x: usize,
    piece_y: usize,
}

impl Tetris {
    fn new() -> Self {
        let mut board = [[0u8; WIDTH]; HEIGHT];
        let mut rng = rand::thread_rng();
        let idx = rng.gen_range(0..SHAPES.len());
        let current = Piece::new(idx);
        let next = Piece::new(rng.gen_range(0..SHAPES.len()));
        Tetris {
            board,
            score: 0,
            lines: 0,
            level: 1,
            fall_time: 500,
            last_fall: Instant::now(),
            game_over: false,
            paused: false,
            current,
            next,
            piece_x: WIDTH/2 - 1,
            piece_y: 0,
        }
    }

    fn spawn_piece(&mut self) {
        self.current = self.next;
        let mut rng = rand::thread_rng();
        self.next = Piece::new(rng.gen_range(0..SHAPES.len()));
        self.piece_x = WIDTH/2 - self.current.shape[0].len()/2;
        self.piece_y = 0;
        if self.collision(&self.current, self.piece_x, self.piece_y) {
            self.game_over = true;
        }
    }

    fn collision(&self, piece: &Piece, x: usize, y: usize) -> bool {
        for row in 0..piece.shape.len() {
            for col in 0..piece.shape[row].len() {
                if piece.shape[row][col] == 1 {
                    let bx = x + col;
                    let by = y + row;
                    if bx >= WIDTH || by >= HEIGHT {
                        return true;
                    }
                    if by < HEIGHT && self.board[by][bx] != 0 {
                        return true;
                    }
                }
            }
        }
        false
    }

    fn lock_piece(&mut self) {
        for row in 0..self.current.shape.len() {
            for col in 0..self.current.shape[row].len() {
                if self.current.shape[row][col] == 1 {
                    let by = self.piece_y + row;
                    let bx = self.piece_x + col;
                    if by < HEIGHT {
                        self.board[by][bx] = 1;
                    }
                }
            }
        }
        self.clear_lines();
        self.spawn_piece();
    }

    fn clear_lines(&mut self) {
        let mut cleared = 0;
        let mut y = HEIGHT as i32 - 1;
        while y >= 0 {
            let mut full = true;
            for x in 0..WIDTH {
                if self.board[y as usize][x] == 0 {
                    full = false;
                    break;
                }
            }
            if full {
                // сдвигаем строки вниз
                for ny in (1..=y as usize).rev() {
                    self.board[ny] = self.board[ny-1];
                }
                self.board[0] = [0; WIDTH];
                cleared += 1;
                // остаемся на той же строке
                y += 1;
            }
            y -= 1;
        }
        if cleared > 0 {
            self.lines += cleared;
            self.score += cleared * 100;
            self.level = self.lines / 10 + 1;
            self.fall_time = std::cmp::max(100, 500 - (self.level - 1) * 30);
        }
    }

    fn move_piece(&mut self, dx: i32, dy: i32) {
        let nx = (self.piece_x as i32 + dx) as usize;
        let ny = (self.piece_y as i32 + dy) as usize;
        if !self.collision(&self.current, nx, ny) {
            self.piece_x = nx;
            self.piece_y = ny;
            return;
        }
        if dy == 1 {
            self.lock_piece();
        }
    }

    fn rotate(&mut self) {
        // поворот матрицы
        let rows = self.current.shape.len();
        let cols = self.current.shape[0].len();
        let mut rotated = vec![vec![0u8; rows]; cols];
        for i in 0..rows {
            for j in 0..cols {
                rotated[j][rows-1-i] = self.current.shape[i][j];
            }
        }
        let new_piece = Piece { shape: rotated, color: self.current.color };
        if !self.collision(&new_piece, self.piece_x, self.piece_y) {
            self.current = new_piece;
        } else {
            // wallkick - пробуем сдвинуть
            for dx in vec![-1, 1] {
                let nx = (self.piece_x as i32 + dx) as usize;
                if !self.collision(&new_piece, nx, self.piece_y) {
                    self.piece_x = nx;
                    self.current = new_piece;
                    break;
                }
            }
        }
    }

    fn drop(&mut self) {
        while !self.collision(&self.current, self.piece_x, self.piece_y + 1) {
            self.piece_y += 1;
        }
        self.lock_piece();
    }

    fn draw(&self) {
        let mut output = String::new();
        // очистка
        output.push_str("\x1b[2J\x1b[H");
        // рамка
        output.push_str("+");
        for _ in 0..WIDTH { output.push('-'); }
        output.push_str("+\n");
        for y in 0..HEIGHT {
            output.push('|');
            for x in 0..WIDTH {
                let mut ch = ' ';
                let mut color = 0;
                // поле
                if self.board[y][x] != 0 {
                    ch = '█';
                    color = 33 + (x % 6) as u16;
                }
                // текущая фигура
                for row in 0..self.current.shape.len() {
                    for col in 0..self.current.shape[row].len() {
                        if self.current.shape[row][col] == 1 {
                            let by = self.piece_y + row;
                            let bx = self.piece_x + col;
                            if by == y && bx == x {
                                ch = '█';
                                color = self.current.color;
                            }
                        }
                    }
                }
                if ch != ' ' {
                    output.push_str(&format!("\x1b[38;5;{}m{}\x1b[0m", color, ch));
                } else {
                    output.push(' ');
                }
            }
            output.push_str("|\n");
        }
        output.push_str("+");
        for _ in 0..WIDTH { output.push('-'); }
        output.push_str("+\n");
        // info
        output.push_str(&format!("Score: {}  Level: {}  Lines: {}\n", self.score, self.level, self.lines));
        if self.paused { output.push_str("PAUSED\n"); }
        if self.game_over { output.push_str("GAME OVER\n"); }
        print!("{}", output);
    }

    fn run(&mut self) {
        let mut stdout = stdout();
        execute!(stdout, terminal::EnterAlternateScreen, cursor::Hide).unwrap();
        terminal::enable_raw_mode().unwrap();

        let mut tick_counter = 0;
        loop {
            self.draw();

            if self.game_over {
                break;
            }

            // обработка ввода
            if event::poll(Duration::from_millis(20)).unwrap() {
                if let Event::Key(key) = event::read().unwrap() {
                    if key.kind == KeyEventKind::Press {
                        match key.code {
                            KeyCode::Char('q') => break,
                            KeyCode::Char('p') => self.paused = !self.paused,
                            _ if !self.paused => {
                                match key.code {
                                    KeyCode::Left => self.move_piece(-1, 0),
                                    KeyCode::Right => self.move_piece(1, 0),
                                    KeyCode::Down => self.move_piece(0, 1),
                                    KeyCode::Up => self.rotate(),
                                    KeyCode::Char(' ') => self.drop(),
                                    _ => {}
                                }
                            }
                            _ => {}
                        }
                    }
                }
            }

            // падение
            let now = Instant::now();
            if !self.paused && now.duration_since(self.last_fall) >= Duration::from_millis(self.fall_time) {
                self.move_piece(0, 1);
                self.last_fall = now;
            }

            std::thread::sleep(Duration::from_millis(20));
        }

        terminal::disable_raw_mode().unwrap();
        execute!(stdout, terminal::LeaveAlternateScreen, cursor::Show).unwrap();
    }
}

fn main() {
    let mut game = Tetris::new();
    game.run();
}
